{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'svg'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "try:\n",
    "    from IPython.display import SVG, clear_output\n",
    "    import numpy as np\n",
    "    import msprime as msprime\n",
    "    import matplotlib.pyplot as plt\n",
    "    import statsmodels.api as sm \n",
    "    \n",
    "except ModuleNotFoundError:\n",
    "    try: \n",
    "        print('ERROR: missing required packages.')\n",
    "        install = input('Attempt to install required modules from requirements.txt? [Y/N] ').lower()\n",
    "        if install == 'y':\n",
    "            os.system(\"conda install --file ./requirements.txt\")\n",
    "        else:\n",
    "            print(\"EXITING\")\n",
    "            exit()\n",
    "\n",
    "    except Exception as e: \n",
    "        print(\"ERROR: failed installing modules.\")\n",
    "        print(e)\n",
    "        print('\\nEXITING')\n",
    "        exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "demography = msprime.Demography()\n",
    "#demography.add_population(name=\"A\", initial_size=7e9, growth_rate=0.006729)    \n",
    "#demography.add_population(name=\"A\", initial_size=1, growth_rate=0)   \n",
    "\n",
    "global s_errors_live\n",
    "global N_range_live\n",
    "global threshold_exceed_live\n",
    "s_errors_live = []\n",
    "N_range_live = []\n",
    "threshold_exceed_live = []\n",
    "\n",
    "def ensemble(N):\n",
    "    # given pop. size N, determines number of ensembles needed\n",
    "    # for trade-off of sim. time and accuracy.\n",
    "    no_ensembles = np.arange(1,1+int(5**(4-N)))\n",
    "    if len(no_ensembles) > 1:\n",
    "        return no_ensembles\n",
    "    else: \n",
    "        return [1,2]\n",
    "\n",
    "\n",
    "def produce_reg_var(N, beta, error, sig_SNP, causal=False, *args, **kwargs):\n",
    "    # return a list of matrices needed for OLS regression\n",
    "    # causal cn be set to 'all', 'causal', 'noncausal' depending on which SNPs are desired\n",
    "    reg_vars = []\n",
    "    X_red = []      # entries are columns of X, dimensions of N x 1\n",
    "    beta_red = []       # entries are effect size for each site\n",
    "    Y_red = []      # entries are Y for each SNP\n",
    "\n",
    "    # iterate over the SNPs that are significant\n",
    "    for j, SNP in enumerate(sig_SNP):\n",
    "        SNP_beta = beta[int(SNP[0])]\n",
    "        genotype = SNP[1]\n",
    "        # keeps certain SNPs based on preference for causal etc.\n",
    "        if not causal or causal == 'all':\n",
    "            X_red.append(genotype)\n",
    "            beta_red.append(SNP_beta)\n",
    "            Y_red.append(genotype*SNP_beta+error) \n",
    "        if causal == 'causal' and not SNP_beta == 0:\n",
    "            X_red.append(genotype)\n",
    "            beta_red.append(SNP_beta)\n",
    "            Y_red.append(genotype*SNP_beta+error)\n",
    "        if causal == 'noncausal' and SNP_beta == 0:\n",
    "            X_red.append(genotype)\n",
    "            beta_red.append(SNP_beta)\n",
    "            Y_red.append(genotype*SNP_beta+error)\n",
    "            \n",
    "    beta_red = np.array(beta_red)       # list -> array\n",
    "\n",
    "    # add matrices for this seed to greater list of regression variables\n",
    "    reg_vars.append([X_red, beta_red, Y_red])\n",
    "    return reg_vars\n",
    "\n",
    "\n",
    "class AncestrySimulations():\n",
    "    '''For generating trees and performing statistical analysis for given population exponent N, and demography.'''\n",
    "    def __init__(\n",
    "        self, \n",
    "        pop_exp, \n",
    "        model=None, \n",
    "        demography=None, \n",
    "        sequence_length=int(10000), \n",
    "        heritability=0.8, \n",
    "        recombination=False, \n",
    "        mutation=[True,1.1e-3], \n",
    "        restrict_SNPs=[True,0.95],\n",
    "        multi_seed=True, \n",
    "        *args, \n",
    "        **kwargs\n",
    "        ):\n",
    "        \n",
    "        # creates percentage complete counter\n",
    "        try:\n",
    "            print(str(round((pop_exp-pop_exp_range[0])/(max(pop_exp_range)-min(pop_exp_range))*100, 2)) + '% complete') \n",
    "        except NameError as e:\n",
    "            print(e)\n",
    "            pass\n",
    "        \n",
    "        mutation, rate = mutation[0], mutation[1]\n",
    "        self.mutation = mutation\n",
    "        # population size is scaled exponentially to cut number of simulations\n",
    "        self.N = int(10**pop_exp)\n",
    "        \n",
    "        # sets whether to iterate over multiple seeds or not \n",
    "        # generates seeds needed to perform ensemble average of tree data\n",
    "        if multi_seed == True:\n",
    "            seeds = ensemble(pop_exp)\n",
    "            print('N: {}\\nNumber of seeds for this N: {}.'.format(self.N, len(seeds)))\n",
    "        if multi_seed == False: \n",
    "            seeds = [1]\n",
    "\n",
    "        sig_g = 5.2/np.sqrt(self.N)   # empirical variance of XÎ² \n",
    "        h = heritability**2\n",
    "        \n",
    "        np.random.seed(0)\n",
    "        self.recombination_rate = recombination\n",
    "        self.sequence_length = sequence_length\n",
    "\n",
    "        self.is_causal = np.random.uniform(0, 1, sequence_length) > restrict_SNPs[1]\n",
    "        self.beta = np.random.normal(sig_g, 1, sequence_length) * self.is_causal\n",
    "        self.error = np.random.normal(0, (sig_g**2)*((1/(h))-1), self.N)\n",
    "\n",
    "        self.all_sig_SNP=[]\n",
    "        \n",
    "        # iterate over seeds and generate trees for each\n",
    "        for seed in seeds:\n",
    "            ts = msprime.sim_ancestry(\n",
    "            self.N, \n",
    "            model=None,\n",
    "            ploidy=1,\n",
    "            demography=demography,\n",
    "            sequence_length=self.sequence_length,\n",
    "            random_seed=seed,\n",
    "            recombination_rate=recombination  \n",
    "            )\n",
    "\n",
    "            if mutation == True:\n",
    "                self.mts = msprime.sim_mutations(\n",
    "                    ts, \n",
    "                    model = msprime.BinaryMutationModel(),\n",
    "                    rate=rate,    #WAS 1.1e-3\n",
    "                    random_seed=seed\n",
    "                    )  \n",
    "            else:\n",
    "                self.mts = ts\n",
    "\n",
    "            # limit for how many N need SNP for it to be considered significant, 1% here\n",
    "            if restrict_SNPs[0] == True:\n",
    "                self.sig_SNP_threshold = self.N/100\n",
    "            else:\n",
    "                self.sig_SNP_threshold = 0\n",
    "\n",
    "            # form of a SNP is (site position, who has mutation)        \n",
    "            self.sig_SNP = [(var.site.position, var.genotypes) for var in self.mts.variants() if len(*np.nonzero(var.genotypes)) > self.sig_SNP_threshold]\n",
    "            self.all_sig_SNP.append(self.sig_SNP)\n",
    "           \n",
    "    # generates the standard error of the estimates of beta and plots them live    \n",
    "    def s_e(self, causal='all', live_plot=False):\n",
    "        for sig_SNP in self.all_sig_SNP:\n",
    "            s_error_arr = [] \n",
    "            # list of the matrices needed to carry out regression, reduced to only significant SNPs\n",
    "            reg_vars = produce_reg_var(self.N, self.beta, self.error, sig_SNP, causal=causal) \n",
    "            s_error_seed = np.zeros(len(reg_vars))\n",
    "            # iteration over the lists in greater regression varibales matrix \n",
    "            for count, data in enumerate(reg_vars):\n",
    "                Y = data[2]\n",
    "                X = data[0]\n",
    "                total_beta_error = np.zeros(len(X))\n",
    "                # ordinary least squares regression on Y, X\n",
    "                for j, X_j in enumerate(X):\n",
    "                # ordinary least squares regression on Y, X\n",
    "                    ols = sm.OLS(np.array(Y[j]), np.array(X_j))\n",
    "                    ols_result = ols.fit()\n",
    "                    total_beta_error[j] = ols_result.bse**0.5\n",
    "                s_error_seed[count] = np.mean(total_beta_error)\n",
    "                s_error_arr.append(s_error_seed)\n",
    "            \n",
    "       \n",
    "        # standard error for this N is mean of all the seeds' standard errors\n",
    "        s_error = np.mean(s_error_arr)\n",
    "        if live_plot == True:\n",
    "            fig1, live_ax = plt.subplots()\n",
    "            live_ax.set_xlabel('N')\n",
    "            live_ax.set_ylabel(r's.e. of $\\^\\beta$')\n",
    "            live_ax.set_title(r'Stadard error of estimated $\\^\\beta$')\n",
    "\n",
    "            # deals with anomalous s_error (can happen for low N if there are an insufficient number of seeds)\n",
    "            if not s_error == float(\"inf\"):\n",
    "                N_range_live.append(self.N)\n",
    "                if s_error > 1000:\n",
    "                    print('s.e. above threshold: {}'.format(s_error))\n",
    "                    s_errors_live.append(s_errors_live[-1])\n",
    "                    threshold_exceed_live.append(s_errors_live[-1])\n",
    "                else: \n",
    "                    s_errors_live.append(s_error)\n",
    "                    threshold_exceed_live.append(None)\n",
    "\n",
    "\n",
    "            live_ax.plot(N_range_live, s_errors_live)\n",
    "            live_ax.plot(N_range_live, threshold_exceed_live, 'rx')\n",
    "            live_ax.grid()\n",
    "\n",
    "            clear_output(wait=True)\n",
    "            plt.show() \n",
    "        return s_error\n",
    "\n",
    "    # plots the effect size of a give SNP against its position on the genome\n",
    "    def SNP_plot(self, causal='all'):\n",
    "        # produce regression variables\n",
    "        reg_vars = produce_reg_var(self.N, self.beta, self.error, self.sig_SNP, causal=causal)        \n",
    "        Y = reg_vars[0][2]\n",
    "        X = reg_vars[0][0]\n",
    "        beta_est = np.zeros(len(X))\n",
    "\n",
    "        for j in range(X.shape[0]):\n",
    "            X_j = X[j][:]\n",
    "            # ordinary least squares regression on Y, X\n",
    "            ols = sm.OLS(np.array([Y[j]]), np.array(X_j))\n",
    "            # extract estimated beta\n",
    "            beta_est[j] = ols.fit().params\n",
    "        \n",
    "        # recover full length matrix of significant SNPs\n",
    "        sig_locs = []\n",
    "        full_beta_est = []\n",
    "        \n",
    "        beta_actual = np.zeros(self.sequence_length)\n",
    "        \n",
    "        for count, data in enumerate(self.sig_SNP):\n",
    "            if abs(beta_est[count]) > 1e-14:\n",
    "                sig_locs.append(int(data[0]))\n",
    "                full_beta_est.append(abs(beta_est[count]))\n",
    "            beta_actual[int(data[0])] = abs(reg_vars[0][1][count])\n",
    "\n",
    "            \n",
    "        SNP_fig, ax = plt.subplots()\n",
    "        ax.set_xlabel('Site')\n",
    "        ax.set_ylabel('Effect size')\n",
    "        ax.set_title('Effect size of SNP at each location,\\nrecombination rate {}'.format(self.recombination_rate))\n",
    "        ax.plot(sig_locs, full_beta_est, 'rx', label='Estimated')\n",
    "        ax.plot(np.arange(self.sequence_length), beta_actual, 'g:', label='Actual')\n",
    "        ax.legend()\n",
    "        return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "global pop_exp_range\n",
    "# the range of exponents used to generate N = 10**(pop_exp)\n",
    "pop_exp_range = np.linspace(2.5,5,50)\n",
    "\n",
    "s_errors_0_8 = np.zeros(len(pop_exp_range))\n",
    "s_errors_0_1 = np.zeros(len(pop_exp_range))\n",
    "s_errors_0_99 = np.zeros(len(pop_exp_range))\n",
    "s_errors_0_01 = np.zeros(len(pop_exp_range))\n",
    "\n",
    "for count, N in enumerate(pop_exp_range):\n",
    "    s_errors_0_8[count] = AncestrySimulations(\n",
    "        N, \n",
    "        model='dtwf',\n",
    "        mutation=[True,1e-3]\n",
    "        ).s_e(\n",
    "            causal='all'\n",
    "            )\n",
    "\n",
    "    s_errors_0_1[count] = AncestrySimulations(\n",
    "        N, \n",
    "        model='dtwf',\n",
    "        heritability=0.1,\n",
    "        mutation=[True,1e-3]\n",
    "        ).s_e(\n",
    "            causal='all'\n",
    "            )\n",
    "\n",
    "    s_errors_0_99[count] = AncestrySimulations(\n",
    "        N, \n",
    "        model='dtwf',\n",
    "        heritability=0.99,\n",
    "        mutation=[True,1e-3]\n",
    "        ).s_e(\n",
    "            causal='all'\n",
    "            )\n",
    "\n",
    "    s_errors_0_01[count] = AncestrySimulations(\n",
    "        N, \n",
    "        model='dtwf',\n",
    "        heritability=0.01,\n",
    "        mutation=[True,1e-3]\n",
    "        ).s_e(\n",
    "            causal='all'\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.set_xlabel('log(N)')\n",
    "ax.set_ylabel(r'log(s.e.($\\^\\beta))$')\n",
    "ax.set_title(r'log(s.e.($\\^\\beta)) \\, vs. \\, log(N)$')\n",
    "ax.set_xscale(\"log\")\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_ylim([10e-5,10e-1])\n",
    "ax.grid()\n",
    "ax.plot(10**pop_exp_range, s_errors_0_01, 'm', label='0.01')\n",
    "ax.plot(10**pop_exp_range, s_errors_0_1, 'r', label='0.1')\n",
    "ax.plot(10**pop_exp_range, s_errors_0_8, label='0.8')\n",
    "ax.plot(10**pop_exp_range, s_errors_0_99, 'g', label='0.99')\n",
    "ax.legend(title='$h^{2}$')\n",
    "\n",
    "fig, ax2 = plt.subplots()\n",
    "ax2.set_xlabel('N')\n",
    "ax2.set_ylabel(r's.e.($\\^\\beta)$')\n",
    "ax2.set_title(r's.e.($\\^\\beta) \\, vs. \\, N$')\n",
    "ax2.grid()\n",
    "ax2.plot(10**pop_exp_range, s_errors_0_01, 'm', label='0.01')\n",
    "ax2.plot(10**pop_exp_range, s_errors_0_1,'r', label='0.1')\n",
    "ax2.plot(10**pop_exp_range, s_errors_0_8, label='0.8')\n",
    "ax2.plot(10**pop_exp_range, s_errors_0_99, 'g', label='0.99')\n",
    "ax2.legend(title='$h^{2}$')\n",
    "\n",
    "print((np.log(s_errors_0_8[-1])-np.log(s_errors_0_8[0]))/(np.log(100000)-np.log(316)))"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c69a9997089f27056cf59ddd46ecd0608f83a003c009680939ddabbbe7d0b9c3"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 ('msprime-env')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
